Improvements for Load Balanced HA Master Deployments

dgoodwin opened the pull request:
Explicitly set all "nodes" schedulable when there are no dedicated nodes. (only systems running master+node)

Warn the user to add dedicated nodes if they have met the min requirements for an HA deployment.

Consider specifying 2 masters an error condition for both attended and unattended installs. You must have 3 for etcd clustering to work correctly.

Allow more than 3 masters for HA installations if desired.

Block re-use of a master or node as the load balancer in attended install. (looks like it was prevented in unattended, but not attended)

Remove prompts to continue in unattanded installs. 

Added various tests for unattended HA installs.

Some pylint refactoring to break down test module.

Added an alternate summary of hosts entered so far during an interactive install.

Revamped some of the text printed during interactive install.

dgoodwin commented:
Forgot to ping @brenton and @smunilla

detiber commented on utils/src/ooinstall/cli_installer.py:
```diff
@@ -261,23 +307,32 @@ def check_hosts_config(oo_cfg):
-of your choice to balance the master API (port 8443) on all master hosts.
+WARNING: No master load balancer specified in config. If you proceed you will
+need to provide a load balancing solution of your choice to balance the
+API (port 8443) on all master hosts.
```
We should be sure to inform them in the prompt that the load balancer needs to be configured before continuing installation

dgoodwin replied:
In a bring your own load balancer scenario do we actually need to know it? I don't think we do anything with that information.

In interactive install it'll show this message and ask you to confirm to continue, in unattended it'll just do it. Should either of those change?

brenton replied:
If someone brings their own loadbalancer we have to mint certificates that list that name as a Subject Alternative Name.

dgoodwin replied:
Aha so they *must* tell us about a LB, I will adjust, thanks guys.

detiber replied:
We need to know the fqdn used for the load balancer URL in order to add it to the generated kubeconfigs and certificates.

If the load balancer isn't pre-configured, then the later steps that try to contact the master will fail for being unable to reach the master(s).

detiber replied:
Well, technically, they could specify the load-balancer name under openshift_master_named_certificates for the cert part, but we would need to provide a way to specify what we use for the generated kubeconfigs...

It gets even more complicated in cloud based installs, because ideally they would have an internal and an external load balancer...  one to load balance console/api traffic for end users and one for load balancing internal traffic that doesn't traverse the external network.

detiber replied:
To expand on that a bit more...  openshift_master_cluster_hostname would be the one for internal use, openshift_master_cluster_public_hostname for external use.

dgoodwin replied:
Current implementation asks for one hostname, I've updated the text a bit to clarify it's gotta be the FQDN. It uses that one hostname for:

openshift_master_cluster_hostname=node2.aos.example.com
openshift_master_cluster_public_hostname=node2.aos.example.com

Do we want to do a second prompt for an internal FQDN for load balancing internal traffic? (I could have it default to the first FQDN, just press enter to use that) Or is this best left to advanced install?

brenton replied:
Personally I'd prefer the fewest prompts possible.  For internal traffic I would think they could use the openshift or kubernetes Services exposed by the kube-proxy in most cases.  @detiber might know of a reason why someone might want an additional internal load balancer.

detiber replied:
@brenton The tricky part is that the openshift_master_cluster_hostname is the one that is used to put in the node configs for the node to contact the master in the case of an HA deployment, we wouldn't necessarily want that traffic traversing an external security zone.

dgoodwin commented:
Updated for a new warning if HA but with less than 3 dedicated nodes.

Added a new "installation summary", the screen will clear (still scrollable) and print something like the following after each host is added:

*** Installation Summary ***

Hosts:
- m1
  - OpenShift Master
  - OpenShift Node (Unscheduled)
  - Etcd Member
- m2
  - OpenShift Master
  - OpenShift Node (Unscheduled)
  - Etcd Member
- m3
  - OpenShift Master
  - OpenShift Node (Unscheduled)
  - Etcd Member
- n1
  - OpenShift Node (Dedicated)
- n2
  - OpenShift Node (Dedicated)
- n4
  - OpenShift Node (Dedicated)

Total OpenShift Masters: 3
Total OpenShift Nodes: 6

NOTE: Multiple Masters specified, this will be an HA deployment with a separate
etcd cluster. You will be prompted to provide the FQDN of a load balancer once
finished entering hosts.

Do you want to add additional hosts? [y/N]: n


The details for each host and the notes/warnings after will adjust as the config grows and show what will actually happen if they proceed right now a little more clearly.

Also centralized the logic for the details about host schedulability/etcd per TODO in openshift_ansible.py. The host now has methods to check if it should be given separate etcd, or marked unschedulable.

brenton commented:
This is a huge improvement.  I'd say we merge this in if you're ready.

dgoodwin commented:
I think it's ready on my end. Thanks!
